### PR TITLE
Add support for allocating qubits in decompose to cirq.unitary

### PR DESCRIPTION
**Pull request by NoureldinYosri**
This is the simplest implementation to do this in order to unblock cirq-ft. 
This implementation only satisfies the first requirement of https://github.com/quantumlib/Cirq/issues/6101#issuecomment-1568686661
In followup PRs I will address other requirements.

### CODE REVIEW

**Review by tanujkhattar — CHANGES_REQUESTED** [review]
Can you add a test where a `GateA` decomposes into a `GateB` where `GateB` allocates new qubits and acts on 

a) <= 4 qubits
b) > 4 qubits  ?

My guess is that `Case-(b)` above would fail the test right now. The reason is as follows:

1. `_try_decompose_into_operations_and_qubits` internally calls `cirq.decompose_once` for gates / operations etc.
2. `cirq.unitary` uses (1) to get a list of decomposed operations and then delegates the logic of applying the unitary effect of these decompositions to `apply_unitaries` protocol.
https://github.com/quantumlib/Cirq/blob/36d67c15141894ea1a062c5c0a9800f655821fbc/cirq-core/cirq/protocols/unitary_protocol.py#L185

3. `apply_unitary` then acts on each operation in the first-level decomposition and then tries to "apply" it's unitary effect onto the given target tensor. To decide a strategy, it looks at `len(args.axes)`.
https://github.com/quantumlib/Cirq/blob/36d67c15141894ea1a062c5c0a9800f655821fbc/cirq-core/cirq/protocols/apply_unitary_protocol.py#L346-L357

4. Thus, if the unitary operators on <= 4 qubits, the recursive strategy will delegate to `cirq.unitary(OpForGateB)` to get its (narrow) unitary and thus be able to apply the effect of the unitary on the target tensor correctly. 
5. But, if it acts > 4 qubits, then it'll try the `_strat_apply_unitary_from_decompose` which will fail because it currently assumes that decomposition returns an op-tree that's acts only the returned set of `qubits` and thus we'd need to make similar changes to `_strat_apply_unitary_from_decompose` as well. 


I think a better strategy would be to change the return type of `_try_decompose_into_operations_and_qubits` to return `work_qubits` and `ancilla_qubits` instead of only `qubits`. But that would be a bigger change since it would trigger every call site that makes this assumption. We can leave a comment on the issue and track this as an independent improvements (I can also take care of it as part of restructuring the `cirq.decompose` protocol).

File: cirq-core/cirq/protocols/unitary_protocol_test.py [hunk 1]
```diff
@@ -201,6 +235,23 @@ def test_decompose_and_get_unitary():
     np.testing.assert_allclose(_strat_unitary_from_decompose(DummyComposite()), np.eye(1))
     np.testing.assert_allclose(_strat_unitary_from_decompose(OtherComposite()), m2)
 
+    np.testing.assert_allclose(
+        _strat_unitary_from_decompose(GateThatAllocatesAQubit()),
+        GateThatAllocatesAQubit._target_unitary,
+    )
```
Comment (tanujkhattar) [inline cirq-core/cirq/protocols/unitary_protocol_test.py:241]: I would suggest creating a new parameterized test for `test_gate_allocates_new_qubits` and then let all the parameterized gates have a common interface to get the `gate._target_unitary()`. I'd also suggest making `GateThatAllocatesAQubit` a parameterized test where we apply `Z ** self.theta` instead of `Z` to the ancilla. 

I imagine we'd like to cover many different "classes" of gates that can allocate new qubits (eg: clean ancilla, dirty ancilla, recursive allocation on <= 4 qubits, >= 4 qubits etc.)

File: cirq-core/cirq/protocols/unitary_protocol.py [hunk 2]
```diff
@@ -179,15 +179,30 @@ def _strat_unitary_from_decompose(val: Any) -> Optional[np.ndarray]:
...
         return NotImplemented
 
+    all_qubits = frozenset(q for op in operations for q in op.qubits)
+    work_qubits = frozenset(qubits)
+    ancillas = tuple(sorted(q for q in all_qubits if q not in work_qubits))
+
+    ordered_qubits = ancillas + tuple(qubits)
+    val_qid_shape = (2,) * len(
+        ancillas
+    ) + val_qid_shape  # For now ancillas have only one qid_shape = (2,).
```
[the remark this answers is not in the document]
Comment (NoureldinYosri) [reply cirq-core/cirq/protocols/unitary_protocol.py:189]: done

[the remark this answers is not in the document]
Comment (NoureldinYosri) [reply cirq-core/cirq/protocols/unitary_protocol.py:189]: done

**Comment by NoureldinYosri** [comment]
it doesn't matter whether the gate acts on $\leq 4$ or $> 4$ qubits as long as the gate doesn't implement `_apply_unitary_` or `_unitary_`.

the issue isn't `_try_decompose_into_operations_and_qubits` it's that the apply_unitary protocol makes the same assumption as the unitary protocol. that the gates/operations won't allocate new qubits if it decomposes them.

I'm working a fix for the apply unitary protocol, will update this PR soon.

**Comment by NoureldinYosri** [comment]
@tanujkhattar got it. however fixing different places where it's called depends on the logic executed there. that's why the method is not the issue for this PR since we already know that these two protocols don't support ancillas.

The `apply_unitaries` method is  more challenging than the `cirq.unitary` protocol. the reason is that it can recurse/decompose multiple times each time potentially creating new ancillas which won't have place in the buffer allocated in the first call.

cirq.unitary is essentially just convenience method for handling different cases with a fall back to `apply_unitaries` so fixing this method is key.

To that end I have two ideas. but I don't like either of them.

**Comment by tanujkhattar** [comment]
@NoureldinYosri I think you should also look at the approach suggested here: https://github.com/quantumlib/Cirq/issues/6040#issuecomment-1483645721 and implemented as part of https://github.com/quantumlib/Cirq/issues/6081. The two issues seem related.

**Comment by NoureldinYosri** [comment]
@tanujkhattar  the issue for me wasn't keeping track of the new qubits. I have a couple of ways for doing that. The issue is how not to sacrifice the efficiency of the protocol.

The current implementation is not efficient in any way. but it will do the trick & unblock cirq-ft. later I'll improve its efficiency.

btw currently there is no mention of the qubit manager. when we introduce the qubit manager abstraction into cirq we will need to add it here. There is a natural way of doing this by adding the qubit manager as a property of the `args`.

For cirq-ft this is currently not a problem since all decompositions use the global qubit manager and there is no object to pass around.

what do you think?

File: cirq-core/cirq/protocols/apply_unitary_protocol.py [hunk 3]
```diff
@@ -363,6 +374,7 @@ def apply_unitary(
     with warnings.catch_warnings():
         warnings.filterwarnings(action="error", category=np.ComplexWarning)
         for strat in strats:
+            strat = cast(Callable[[Any, ApplyUnitaryArgs], Optional[np.ndarray]], strat)
```
Comment (tanujkhattar) [inline cirq-core/cirq/protocols/apply_unitary_protocol.py:377]: I don't think signature is true for all the strategies? `_strat_apply_unitary_from_apply_unitary` and `_strat_apply_unitary_from_decompose` both don't take the optional matrix parameters. 

Maybe create a helper method for `_strat_apply_unitary_from_unitary` and dispatch to it with the matrix parameter wherever needed?

File: cirq-core/cirq/protocols/apply_unitary_protocol.py [hunk 4]
```diff
@@ -410,17 +422,18 @@ def _strat_apply_unitary_from_apply_unitary(
...
 ) -> Optional[np.ndarray]:
-    # Check for magic method.
-    method = getattr(unitary_value, '_unitary_', None)
-    if method is None:
-        return NotImplemented
+    if matrix is None:
+        # Check for magic method.
+        method = getattr(unitary_value, '_unitary_', None)
+        if method is None:
+            return NotImplemented
```
Comment (tanujkhattar) [inline cirq-core/cirq/protocols/apply_unitary_protocol.py:431]: Why is this needed? Looks like you are calling this method with the `matrix` from `_strat_apply_unitary_from_decompose` after you have already computed the narrowed matrix. Why the fallback on  `_strat_apply_unitary_from_decompose` ?

File: cirq-core/cirq/protocols/apply_unitary_protocol.py [hunk 5]
```diff
@@ -454,7 +467,27 @@ def _strat_apply_unitary_from_decompose(val: Any, args: ApplyUnitaryArgs) -> Opt
...
         return NotImplemented
-    return apply_unitaries(operations, qubits, args, None)
+    all_qubits = frozenset([q for op in operations for q in op.qubits])
+    ancilla = tuple(sorted(all_qubits.difference(qubits)))
+    if not len(ancilla):
+        return apply_unitaries(operations, qubits, args, None)
+    ordered_qubits = ancilla + tuple(qubits)
+    all_qid_shapes = qid_shape_protocol.qid_shape(ordered_qubits)
+    state = qis.eye_tensor(all_qid_shapes, dtype=np.complex128)
+    buffer = np.empty_like(state)
```
Comment (tanujkhattar) [inline cirq-core/cirq/protocols/apply_unitary_protocol.py:477]: I think we should create a `kronecker_product` and `factor` method on `ApplyUnitaryArgs`; similar to `SimulationState`; and then dispatch the logic of allocate new space and extracting out the unitary corresponding to a subspace using function calls to these methods.

File: cirq-core/cirq/protocols/apply_unitary_protocol.py [hunk 6]
```diff
@@ -454,7 +467,27 @@ def _strat_apply_unitary_from_decompose(val: Any, args: ApplyUnitaryArgs) -> Opt
...
+        ApplyUnitaryArgs(state, buffer, range(len(ordered_qubits))),
+        None,
+    )
+    if result is None or result is NotImplemented:
+        return result
+    result = result.reshape((np.prod(all_qid_shapes, dtype=np.int64), -1))
+    val_qid_shape = qid_shape_protocol.qid_shape(qubits)
+    state_vec_length = np.prod(val_qid_shape, dtype=np.int64)
+    result = result[:state_vec_length, :state_vec_length]
+    return _strat_apply_unitary_from_unitary(val, args, matrix=result)
```
Comment (tanujkhattar) [inline cirq-core/cirq/protocols/apply_unitary_protocol.py:490]: Again, not sure why this dispatch to `_strat_apply_unitary_from_unitary` is necessary.

**Review by tanujkhattar — CHANGES_REQUESTED** [review]
Looking good overall. Main suggestion is to extend the `ApplyUnitaryArgs`  class to provide methods to allocate new space for the ancillas and extract out the unitary corresponding to working qubits.

File: cirq-core/cirq/protocols/apply_unitary_protocol.py (hunk 3 above)
Comment (NoureldinYosri) [reply cirq-core/cirq/protocols/apply_unitary_protocol.py:377]: done

**Comment by NoureldinYosri** [comment]
I addressed the comment we had offline. 

I'll keep the code that factors out the unitary as it's for now and I'll add a general `factor_unitary` in a followup PR to cirq-core/cirq/linalg/transformations.py

File: cirq-core/cirq/protocols/apply_unitary_protocol.py [hunk 7]
```diff
@@ -224,6 +224,12 @@ def subspace_index(
             qid_shape=self.target_tensor.shape,
         )
 
+    @classmethod
+    def for_unitary(cls, qid_shapes: Tuple[int, ...]) -> 'ApplyUnitaryArgs':
```
Comment (tanujkhattar) [inline cirq-core/cirq/protocols/apply_unitary_protocol.py:228]: nit: Move the class method up with factory method `default` and add a docstring.

File: cirq-core/cirq/protocols/apply_unitary_protocol.py [hunk 8]
```diff
@@ -274,6 +280,10 @@ def _apply_unitary_(
         """
 
 
+def _strat_apply_unitary_from_unitary_(val: Any, args: ApplyUnitaryArgs) -> Optional[np.ndarray]:
```
Comment (tanujkhattar) [inline cirq-core/cirq/protocols/apply_unitary_protocol.py:283]: Maybe do it the other way so you don't change the interface of `_strat_apply_unitary_from_unitary` and introduce a new `_apply_unitary_from_matrix` which only accepts a `matrix: np.ndarray` and `args: ApplyUnitaryArgs ` and multiplies the `matrix` into `args`. Then both  `_strat_apply_unitary_from_unitary` and `_strat_apply_unitary_from_unitary` will call `_apply_unitary_from_matrix`. This way, you don't need to add the unnecessary check in `_strat_apply_unitary_from_unitary` that the provided `matrix` is None or not.

File: cirq-core/cirq/protocols/apply_unitary_protocol.py [hunk 9]
```diff
@@ -454,7 +465,25 @@ def _strat_apply_unitary_from_decompose(val: Any, args: ApplyUnitaryArgs) -> Opt
...
+    all_qubits = frozenset([q for op in operations for q in op.qubits])
+    ancilla = tuple(sorted(all_qubits.difference(qubits)))
+    if not len(ancilla):
+        return apply_unitaries(operations, qubits, args, None)
+    ordered_qubits = ancilla + tuple(qubits)
+    all_qid_shapes = qid_shape_protocol.qid_shape(ordered_qubits)
+    result = apply_unitaries(
+        operations,
+        ordered_qubits,
+        ApplyUnitaryArgs.for_unitary(qid_shape_protocol.qid_shape(ordered_qubits)),
```
Comment (tanujkhattar) [inline cirq-core/cirq/protocols/apply_unitary_protocol.py:477]:
```suggestion
        ApplyUnitaryArgs.for_unitary(all_qid_shapes),
```

File: cirq-core/cirq/protocols/apply_unitary_protocol.py [hunk 10]
```diff
@@ -454,7 +465,25 @@ def _strat_apply_unitary_from_decompose(val: Any, args: ApplyUnitaryArgs) -> Opt
...
+        None,
+    )
+    if result is None or result is NotImplemented:
+        return result
+    result = result.reshape((np.prod(all_qid_shapes, dtype=np.int64), -1))
+    val_qid_shape = qid_shape_protocol.qid_shape(qubits)
+    state_vec_length = np.prod(val_qid_shape, dtype=np.int64)
+    result = result[:state_vec_length, :state_vec_length]
+    return _strat_apply_unitary_from_unitary(val, args, matrix=result)
```
Comment (tanujkhattar) [inline cirq-core/cirq/protocols/apply_unitary_protocol.py:486]:
```suggestion
    return _apply_unitary_from_matrix(matrix, args)
```

File: cirq-core/cirq/protocols/apply_unitary_protocol.py [hunk 11]
```diff
@@ -410,17 +420,18 @@ def _strat_apply_unitary_from_apply_unitary(
...
+            return NotImplemented
 
-    # Attempt to get the unitary matrix.
-    matrix = method()
-    if matrix is NotImplemented or matrix is None:
-        return matrix
+        # Attempt to get the unitary matrix.
+        matrix = method()
+        if matrix is NotImplemented or matrix is None:
+            return matrix
```
Comment (tanujkhattar) [inline cirq-core/cirq/protocols/apply_unitary_protocol.py:434]:
```suggestion
            return matrix
    return _apply_unitary_from_matrix(matrix, args)
```

**Review by tanujkhattar — APPROVED** [review]
LGTM % minor refactoring nits

File: cirq-core/cirq/testing/sample_gates.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
+# Copyright 2023 The Cirq Developers
```
Comment (tanujkhattar) [inline cirq-core/cirq/testing/sample_gates.py:1]: My proposal for this file was to have a library of well defined gates which cover different cases of ancilla allocation; eg:

1) Allocate ancilla but expect them to be clean and leave them in a clean state (eg: `GateThatAllocatesAQubit`) 
2) Allocate ancilla and expect them to be in any dirty state and leave them in the same state (eg: `GateThatAllocatesTwoQubits`)


For such gates; we should have documentation and tests for their decomposition as part of `cirq-core/cirq/testing/sample_gates_test.py`. The goal is to make these gates reusable for other tests (like reduced unitary / simulations). 

The more simpler cases like `NotDecomposableGate` or more specific cases like `DecomposableGate` which doesn't have any guarantees on the can (and probably should) live in specific test files. 

We can also do this as part of a follow-up PR if you'd prefer; but this was my main motivation for suggesting that we create a separate file in `cirq/testing`

We need to add documentation and tests for gates in this file. The idea is to keep things like `GateThatAllocatesTwoQubits` in this file; which essentially has a decomposition where we can borrow two qubits in any state (with a docstring that clarifies it and a test which ensures that the decomposition is correct).

**Review by tanujkhattar — CHANGES_REQUESTED** [review]
One comment on simple_gates.py -- either we don't create a new file in this PR or we add tests to make sure the new file has coverage.

File: cirq-core/cirq/protocols/apply_unitary_protocol.py (hunk 11 above)
Comment (NoureldinYosri) [reply cirq-core/cirq/protocols/apply_unitary_protocol.py:434]: done

File: cirq-core/cirq/protocols/apply_unitary_protocol.py (hunk 8 above)
Comment (NoureldinYosri) [reply cirq-core/cirq/protocols/apply_unitary_protocol.py:283]: done

File: cirq-core/cirq/testing/sample_gates.py (hunk 12 above)
Comment (NoureldinYosri) [reply cirq-core/cirq/testing/sample_gates.py:1]: Lets keep the interesting gates here as is and I'll improve their docstring and add _test in a followup PR

**Comment by NoureldinYosri** [comment]
@tanujkhattar PTAL

for `testing/sample_gates.py` lets keep it as is for now and I'll improve the classes and add the tests in a follow up PR in which that file is the focus instead of going on a tangent here.

File: cirq-core/cirq/testing/sample_gates.py (hunk 12 above)
Comment (NoureldinYosri) [reply cirq-core/cirq/testing/sample_gates.py:1]: done

**Comment by NoureldinYosri** [comment]
@tanujkhattar  PTAL. I added `sample_gates_test.py` and addressed the other comments.

**Comment by tanujkhattar** [comment]
Here is a commit on top of this PR which suggests a few changes (described below) - https://github.com/tanujkhattar/Cirq/commit/14dd571d64d54b6aca18bb43d9871f4db6c976c2


The tests for `sample_gates.py` right now test a) the target given unitary as hardcoded in the original class and b) the reduced unitary assuming the `cirq.unitary` implementation is correct. Ideally, we should the tests should verify that the decomposition is correct (i.e. have an assertion on the structure of the decomposed unitary). 

I've refactored the `sample_gates.py` file so that it now contains two gates - `PhaseUsingCleanAncilla` and `PhaseUsingDirtyAncilla` each of which apply a phase to a given basis state using clean / dirty ancillas. 

I've also fixed a few styling nits like added docstrings and modified function names. Can you please incorporate the changes from my commit into your PR and then we can go ahead and merge?

**Comment by NoureldinYosri** [comment]
done. one thing I noticed is that the `Isolated pytest Ubuntu` test is flaky, it sometimes fails on unrelated tests but rerunning it makes it succeed.

**Comment by shef4** [comment]
Hi, I set up my dev enviroment to start working on this issue. Silly question but how do i idetify which protcols arent workign ciorrect and need support. is there a test case i can run or a post with specfic details?

**Comment by NoureldinYosri** [comment]
@shef4 sounds good, thanks for your help

In the mean time I will soon add the consistancy and correctness checks.

**Comment by shef4** [comment]
sure, thing! I looked into the problem more and wanted to get some input before coding. 

From what I understood from the comments and the code files. I think it makes sense to add a simple test for "earsing entagled qubits using measurment" in apply_unitary_protocol_test.py that utilizes the apply_unitaries function. 

The apply_unitaries function currently terminates if one of the the given unitary values does not have a unitary effect ,but does not rollback changes from previous unitray values. 

A bruteforce approach could be to apply all unitary operations and check at the end if the work qubit is still unitary. This would probaly require the least amount of change to the code, but would make apply_unitaries less effecient. I could also add a separate function.

Just to clearify, can the work qubit be non-unitary during the operation sequence as long as it is unitary at the end?

**Comment by NoureldinYosri** [comment]
> A bruteforce approach could be to apply all unitary operations and check at the end if the work qubit is still unitary. 

@shef4 the product of unitaries is always a unitary.

the problem we have is different. if some operation is not unitary and in that case there is no `_apply_unitary_` or `_unitary_` methods by definition. however multiple non unitary operations could lead to a unitary operation. take for example the [CCNOT gate](https://en.wikipedia.org/wiki/Toffoli_gate)  which flips the last qubit if and only if the first 2 are ones. its unitary is given by

$$
\begin{bmatrix}
1 & 0 & 0 & 0 & 0 & 0 & 0 & 0\\
 0 & 1 & 0 & 0 & 0 & 0 & 0 & 0\\
 0 & 0 & 1 & 0 & 0 & 0 & 0 & 0\\
 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0\\
 0 & 0 & 0 & 0 & 1 & 0 & 0 & 0\\
 0 & 0 & 0 & 0 & 0 & 1 & 0 & 0\\
 0 & 0 & 0 & 0 & 0 & 0 & 0 & 1\\
 0 & 0 & 0 & 0 & 0 & 0 & 1 & 0\\
\end{bmatrix}
$$

This gate can be implemented using [And](https://github.com/quantumlib/Cirq/blob/master/cirq-ft/cirq_ft/algos/and_gate.py#L24)  gate as

```py3

class CCNOTUsingAnd(cirq_ft.GateWithRegisters):

    @cached_property   
    def registers(self):
        return cirq_ft.Registers.build(control=2, target=1)

    def decompose_from_registers(
        self, *, context: cirq.DecompositionContext, **quregs: Sequence[cirq.Qid]
    ) -> cirq.OP_TREE:
        control, target = quregs['control'], quregs['target']
        (ancilla,) = context.qubit_manager.qalloc(1)

        yield cirq_ft.And().on(*control, ancilla)
        yield cirq.CNOT(ancilla, t)
        yield cirq_ft.And(adjoint=True)(c0, c1, ancilla)
```

so task 2 is to support something like
```py3
u = cirq.unitary(CCNOTUsingAnd)   # Must succeeds to build the unitary.
assert np.allclose(u, cirq.unitary(cirq.CCNOT))   # Must be the same as the unitary of a CCNOT
```

currently the line `cirq.unitary(CCNOTUsingAnd)` would fail since `And(adjoint=True)` is not a unitary operation since its decomposition contains a [measurement](https://github.com/quantumlib/Cirq/blob/master/cirq-ft/cirq_ft/algos/and_gate.py#L99) . the purpose of task 2 is support this and simillar use cases.

**Comment by shef4** [comment]
Thanks for the clarification, there was a typo in my brute-force solution comment. 

what I meant was that from my understanding, `cirq.unitary(CCNOTUsingAnd)` call's `_unitary_from_decompose()` under the hood to evaluate `CCNOTUsingAnd`.

`_unitary_from_decompose()` then calls 'apply_unitaries' to produce a unitary matrix from the `CCNOTUsingAnd`.

assert np.allclose(u, cirq.unitary(cirq.CCNOT))

**Comment by NoureldinYosri** [comment]
@shef4 please check https://github.com/quantumlib/Cirq/issues/6182
